CORE-13360 admin: implement generic AIP-160 filter parsing

pgellert opened the pull request:
As part of the client monitoring project, we are going to adopt (a subset of) the AIP-160 spec to allow for complex filtering of resources returned by `List*` endpoints of the Admin v2 API.

This PR implements a parser that can be used to parse AIP-160 compliant, human-readable filter expression strings into filter_predicate's to allow filtering protobuf objects.

Supports a subset of the AIP-160 spec:
- Field comparisons (`=`, `!=`, `<`, `>`, `<=`, `>=`)
- Logical AND chaining: condition1 AND condition2
- Nested field access: parent.child = value
- Escape sequences: field = "string with \"quotes\""
- Enum types
- RFC3339 timestamps and ISO-like duration parsing via abseil

Ref: https://google.aip.dev/160

Fixes https://redpandadata.atlassian.net/browse/CORE-13360

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
pbgen: lookup_fields take const span

This allows us to pass in a `const std::vector<int32_t>&` to
`lookup_fields`.

pgellert committed:
admin: implement field registry for filtering

Introduce an interface for a field registry that allows looking up a set
of fields by name.

This allows us to:
 * encapsulate the logic of translating a field name to a field accessor
 * decouple the filtering implementation from the current redpanda
   protobuf serde implementation
 * reuse the field registry later for the implementation of sorting
 * mock the interface for testing

pgellert committed:
admin: implement generic AIP-160 filter parsing

This implements a parser that can be used to parse AIP-160 compliant,
human-readable filter expression strings into filter_predicate's to
allow filtering protobuf objects.

Supports a subset of the AIP-160 spec:
- Field comparisons (`=`, `!=`, `<`, `>`, `<=`, `>=`)
- Logical AND chaining: condition1 AND condition2
- Nested field access: parent.child = value
- Escape sequences: field = "string with \"quotes\""
- Enum types
- RFC3339 timestamps and ISO-like duration parsing via abseil

Ref: https://google.aip.dev/160

pgellert committed:
admin/tests: test the AIP-160 filter parser logic

rockwotj commented on bazel/pbgen/main.go:
Please run

```
bazel run //src/v/serde/protobuf/tests:protogen_golden_conformance_test_update
```

rockwotj commented on src/v/redpanda/admin/BUILD:
```diff
@@ -143,6 +143,22 @@ redpanda_cc_library(
+    ],
+    visibility = ["//visibility:public"],
+    deps = [
+        "//src/v/redpanda/admin/tests:aip_filter_test_messages_redpanda_proto",
```
Does this proto exist in commit 3a10a5b8da6e7387d587accd8aa294ffe01a9247

rockwotj commented on src/v/redpanda/admin/field_registry.h:
```diff
@@ -0,0 +1,94 @@
+template<typename T>
+struct field_accessor_info {
+    // clang-format off
+    using int64_getter = named_type<std::function<int64_t(T&)>, struct int64_getter_tag>;
```
should these be std::function or ss::noncopyable_function or even absl::FunctionRef?

rockwotj commented on src/v/redpanda/admin/field_registry.cc:
```diff
@@ -0,0 +1,148 @@
+    return std::make_unique<protobuf_field_registry<T>>();
+}
+
+// Explicitly instantiating for all supported protobuf types to avoid having to
+// expose all the templated code in the header
+template class protobuf_field_registry<aip_filter_test::test_message>;
+template std::unique_ptr<field_registry<aip_filter_test::test_message>>
+make_field_registry<aip_filter_test::test_message>();
```
this feels a bit strange pattern. Shouldn't instead we do something like replace `make_field_registry` in the header with `make_aip_filter_test_field_registry`?

rockwotj commented on src/v/redpanda/admin/field_registry.cc:
```diff
@@ -0,0 +1,148 @@
+        throw std::invalid_argument("Invalid field path: " + field_path);
+    }
+
+    T sample_instance;
+    auto field_opt = sample_instance.lookup_field(*field_numbers_opt);
+    if (!field_opt) {
+        throw std::invalid_argument("Field not found: " + field_path);
+    }
```
this is not needed, because `convert_field_path_to_numbers` will never return a number list without this method also always returning a value

rockwotj commented on src/v/redpanda/admin/field_registry.h:
```diff
@@ -0,0 +1,94 @@
+template<typename T>
+concept RedpandaProtobufMessage = std::derived_from<T, serde::pb::base_message>;
+
+/// A registry implementation that uses reflection to dynamically look up fields
+/// on a protobuf message generated using redpanda's protobuf serde generator.
+template<RedpandaProtobufMessage T>
+class protobuf_field_registry : public field_registry<T> {
```
I'm not really yet convinced this class and stuff buys us anything useful, because it seems to me like the lambda wrappers just are there to cache the field path to field number conversion.

rockwotj commented on src/v/redpanda/admin/aip_filter.cc:
```diff
@@ -0,0 +1,371 @@
+namespace admin {
+
+namespace spirit_parser {
+namespace x3 = boost::spirit::x3;
```
should we be concerned about this that is no longer maintained? https://github.com/boostorg/spirit/ I worry about relying on esoteric boost libraries. Looks like v4 is not yet ready for usage

Because I cannot really reason about the grammar stuff below. Like maybe we should consider https://github.com/foonathan/lexy?

rockwotj reviewed (commented):
There are two big points we need to talk over:

1. Usage of boost::spirit, I'm a bit concern the library is unmaintained. I'm vaguely familar with [lexy](https://github.com/foonathan/lexy), and that seems to be actively developed and I would be more comfortable with adding a dependency on. 

In iceberg for parsing out the expressions for partition specs, we just wrote a small handrolled parser: https://github.com/redpanda-data/redpanda/blob/dev/src/v/datalake/partition_spec_parser.cc




2. There are a lot of templates here. I think we can simplify the implementation a lot with some type erasure. Basically think you could have an API like this and it would be much simpler:

```c++

struct aip_filter_config {
  // A getter for converting field numbers -> a value_t
  ss::noncopyable_function<serde::pb::field::value_t(serde::pb::base_message*, std::span<const int32_t> field_numbers)> field_value_getter;
  // A converter from field path in filter expression to field numbers. 
  // If std::nullopt is returned the filter specified a path that does 
  // not exist and filter creation will throw an exception.
  ss::noncopyable_function<std::optional<std::vector<int32_t>>(ss::string field_path)> field_path_converter;
  // The user specified filter expression.
  ss::sstring filter_expression;
}

// Create a filter function from the specified configuration. 
// This method will throw if the filter expression is invalid
// either syntaxically or if it uses a field path that does 
// not exist according to the lookup functions.
// Returns a function that if you pass in the proto message,
// returns true if the message matches the filter.
ss::noncopyable_function<bool(serde::pb::base_message*)> create_aip_filter(aip_filter_config);
```

This interface still enables most of the optimizations around caching the field path -> field number conversion (all the field_path_converter can be called during parsing) but I don't think we get any benefit from the templates for the filter class and the field registry thing. In reality we have some erasure going on already in the field lookup, so I'd ask you to write a benchmark to validate the template version is really worth it performance wise, and if we do that, then I'd still ask for some changes to this PR.

pgellert commented:
Closing this PR because I have reworked a large part of the implementation (spirit -> lexy, no field registry) and moved over to a new PR https://github.com/redpanda-data/redpanda/pull/27704
